src/julia.h

```diff
@@ -1358,6 +1372,50 @@ extern DLLEXPORT jl_compileropts_t jl_compileropts;
 #define JL_COMPILEROPT_DUMPBITCODE_ON 1
 #define JL_COMPILEROPT_DUMPBITCODE_OFF 2
 
+DLLEXPORT void gc_queue_root(void *root);
+void gc_queue_binding(void *bnd);
+void gc_setmark_buf(void *buf, int);
+DLLEXPORT void gc_wb_slow(void* parent, void* ptr);
+
+static inline void gc_wb_binding(void *bnd, void *val)
+{
+    if (__unlikely((*(uintptr_t*)bnd & 1) == 1 && (*(uintptr_t*)val & 1) == 0))
+        gc_queue_binding(bnd);
+}
+
+static inline void gc_wb_fwd(void* parent, void* ptr)
+{
+    // if parent is marked and ptr is clean
+    if(__unlikely((*((uintptr_t*)parent) & 1) == 1 && (*((uintptr_t*)ptr) & 1) == 0)) {
+        // the set lsb indicates this object must stay in the remset until the next
+        // long collection
+        gc_queue_root((void*)((uintptr_t)ptr | 1));
+    }
+}
+
+static inline void gc_wb(void *parent, void *ptr)
+{
+    if (__unlikely((*((uintptr_t*)parent) & 1) == 1 &&
+                   (*((uintptr_t*)ptr) & 1) == 0))
+        gc_queue_root(parent);
+}
+
+static inline void gc_wb_buf(void *parent, void *bufptr)
+{
+    // if parent is marked and buf is not
+    if (__unlikely((*((uintptr_t*)parent) & 1) == 1))
+                   //                   (*((uintptr_t*)bufptr) & 3) != 1))
+          gc_setmark_buf(bufptr, *(uintptr_t*)parent & 3);
```

Comment:
this will mark bufptr (for the next gc pass), but if parent was old, wouldn't this violate the state transitions setup (for the next+1 gc pass) since you have parent(old) pointing to bufptr(new)? (per usage of this function in https://github.com/JuliaLang/julia/blob/9b99e94fdbd48d42b704fcc76466ca06d6abfc92/src/module.c#L86)